### ISSUE TITLE
Need help running Camera-based object detection (3d-yolo, SMOKE) using Python

### ISSUE THREAD

**Issue by zhiheng97** [ApolloAuto/apollo#14982]
We appreciate you go through Apollo documentations and search previous issues before creating an new one. If neither of the sources helped you with your issues, please report the issue using the following form. Please note missing info can delay the response time.

### System information

- Google Colab Notebook
- Python 3.10
- Model files were pulled from Apollo r8.0.0.0 docker image

### Steps to reproduce the issue:

- Load the models using the respective frameworks (3d-yolo -> PaddlePaddle, SMOKE -> PyTorch)
- Load the images from Waymo Open Dataset (Waymo-open-dataset-tf-2-11-0==1.5.1) into tensors
- Run inference 

### Supporting materials (screenshots, command lines, code/script snippets):

- 3d-yolo
  - Here I'm loading 3d-yolo's inference model using PaddlePaddle
  ```
  def initModel(path_to_model: str, model_file_name: str) -> list:
      # Check if paddlepaddle is installed correctly
      paddle.utils.run_check()
  
      # Set the device to cpu
      paddle.set_device('cpu')
  
      paddle.enable_static()
      place = paddle.CPUPlace()
      exe = paddle.static.Executor(place)
      
      [inference_program, feed_target_names, fetch_targets] = paddle.static.load_inference_model(
          path_prefix = path_to_model,
          executor = exe,
          model_filename = model_file_name + '.pdmodel',
          params_filename = model_file_name + '.pdiparams'
      )
      return [inference_program, feed_target_names, fetch_targets, exe]
  
  ```
   - Here I'm loading the images from Waymo Open Dataset as tensors 
 
  ```
  def imageByteToTensor(dataframe, yieldInterval = 20):
    imgs = []
    orignal_img = []
    i = 1
    for _, row in dataframe.iterrows():
      if i % yieldInterval == 0:
        yield [tf.convert_to_tensor(imgs), orignal_img]
      image_dict = v2.CameraImageComponent.from_dict(row)
      image_tensor = tf.image.decode_image(image_dict.image, channels=3, dtype=tf.dtypes.float32)
      image_tensor = tf.image.resize(image_tensor, [800, 1440])
      imgs.append(image_tensor / 255)
      orignal_img.append(image_dict.image)
      i += 1
  ```
  
    - Here I'm running the inference
  ```
  x = imageByteToTensor(images, 1)
  tensor, imgs = next(x)
  res = executor.run(program, feed={target_names[0]: tensor.numpy()}, fetch_list=targets)
  ```
  
    - Here is what the object ```targets``` contain which I presume is the format of ```res```. There's currently no documentation on what these tensors mean, and I've been trying to look for documentation on it. Could any of the developers be kind enough to shed some light? Much appreciated!
  ```
  [var save_infer_model/scale_0 : LOD_TENSOR.shape(-1, 50, 90, 16).dtype(float32).stop_gradient(False), 
  var save_infer_model/scale_1 : LOD_TENSOR.shape(-1, 50, -1, 128).dtype(float32).stop_gradient(False), 
  var save_infer_model/scale_2 : LOD_TENSOR.shape(-1, 50, 90, 32).dtype(float32).stop_gradient(False), 
  var save_infer_model/scale_3 : LOD_TENSOR.shape(-1, 50, 90, 48).dtype(float32).stop_gradient(False), 
  var save_infer_model/scale_4 : LOD_TENSOR.shape(-1, 50, 90, 16).dtype(float32).stop_gradient(False), 
  var save_infer_model/scale_5 : LOD_TENSOR.shape(-1, 50, 90, 16).dtype(float32).stop_gradient(False), 
  var save_infer_model/scale_6 : LOD_TENSOR.shape(-1, 50, 90, 16).dtype(float32).stop_gradient(False), 
  var save_infer_model/scale_7 : LOD_TENSOR.shape(-1, 50, 90, 16).dtype(float32).stop_gradient(False), 
  var save_infer_model/scale_8 : LOD_TENSOR.shape(-1, 50, 90, 16).dtype(float32).stop_gradient(False), 
  var save_infer_model/scale_9 : LOD_TENSOR.shape(-1, 50, 90, 16).dtype(float32).stop_gradient(False), 
  var save_infer_model/scale_10 : LOD_TENSOR.shape(-1, 50, -1, 4).dtype(float32).stop_gradient(False), 
  var save_infer_model/scale_11 : LOD_TENSOR.shape(-1, 50, 90, -1, 4).dtype(float32).stop_gradient(False), 
  var save_infer_model/scale_12 : LOD_TENSOR.shape(-1, 50, 90, 64).dtype(float32).stop_gradient(False), 
  var save_infer_model/scale_13 : LOD_TENSOR.shape(-1, 50, 90, 64).dtype(float32).stop_gradient(False), 
  var save_infer_model/scale_14 : LOD_TENSOR.shape(-1, 64, 200, 360).dtype(float32).stop_gradient(False)]
  ```

- SMOKE
  - Here I'm loading smoke_libtorch_model.pth using PyTorch
  ```
  model = torch.jit.load('/content/drive/MyDrive/fyp/models/smoke_libtorch/smoke_libtorch_model.pth')
  model.eval()
  ```

  - Problem here is that ```model.forward()``` takes in 2 arguments as seen in ```print(model.code())```. So, what should Ks and down_ratios be set to? Much appreciated!
  ```
  def forward(self,
      input: Tensor,
      argument_2: Tuple[Tensor, Tensor]) -> Tuple[Tensor, Tensor]:
    heads = self.heads
    backbone = self.backbone
    Ks, down_ratios, = argument_2
    _0 = (backbone).forward(input, )
    _1 = (heads).forward(_0, Ks, down_ratios, )
    return (_1, _0)
  ```

**Comment by daohu527** [comment]
can i say there are some problems here
```
3d-yolo -> PaddlePaddle   // it's use tensorRT, but not PaddlePaddle
SMOKE -> PyTorch  // it's pytorch version
```

and be sure SMOKE need 3 inputs
https://github.com/ApolloAuto/apollo/blob/446c9c6d481fef9853ab9f167a0e934092058244/modules/perception/camera/lib/obstacle/detector/smoke/smoke_obstacle_detector.cc#L96-L101

You can get more details in 
https://github.com/ApolloAuto/apollo/blob/446c9c6d481fef9853ab9f167a0e934092058244/modules/perception/camera/lib/obstacle/detector/smoke/smoke_obstacle_detector.cc#L313

**Comment by zhiheng97** [comment]
Hi @daohu527,

Thank you for pointing it out and please do correct me as I'm a student who is new to all of this!

I have to apologise for the mix up, I used the 3d-r4 model instead of the 3d-yolo model. 
Based on `apollo_deploy.yml`, I understand that 3d-r4-half is using Caffe framework and has 3 outputs (cls_pred, label, confidence). And there are files `model` and `params` in the docker image that I'm able to load into PaddlePaddle, and hence the `targets` object.
![Screenshot 2023-06-08 at 10 58 20](https://github.com/ApolloAuto/apollo/assets/77670828/1dd77662-247b-4d07-83c7-41717ba107ff)

As for SMOKE, once again thank you for pointing me to where I should be looking at. After connecting some of the dots, I landed on `libtorch_obstacle_detector.cc`, should I use the same values?
[https://github.com/ApolloAuto/apollo/blob/446c9c6d481fef9853ab9f167a0e934092058244/modules/perception/inference/onnx/libtorch_obstacle_detector.cc#L69-L74](https://github.com/ApolloAuto/apollo/blob/446c9c6d481fef9853ab9f167a0e934092058244/modules/perception/inference/onnx/libtorch_obstacle_detector.cc#L69-L74)

I also noticed that values are pulled from a `map`
[https://github.com/ApolloAuto/apollo/blob/446c9c6d481fef9853ab9f167a0e934092058244/modules/perception/inference/onnx/libtorch_obstacle_detector.cc#L110-L125](https://github.com/ApolloAuto/apollo/blob/446c9c6d481fef9853ab9f167a0e934092058244/modules/perception/inference/onnx/libtorch_obstacle_detector.cc#L110-L125)

Thank you!

**Comment by LordonCN** [comment]
I think these will help you

```
    fake_input = torch.zeros([1, 3, 640, 960]).cuda() # apollo use 640 960
    K = torch.tensor([[[1, 0, 0], [0, 1, 0], [0, 0, 1]]]).float().cuda()
    down_ratio = torch.tensor([[10, 8.4375]]).float().cuda()
    cam_info = (K, down_ratio)
    output = model(fake_input, cam_info)
```
